File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/StopObservingBingoEventsHandler.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkSpotActorRequestHandler;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors.PlayerActor;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class StopObservingBingoEventsHandler
        implements ZLinkSpotActorRequestHandler<
                BingoRoomSpot,
                PlayerActor,
                Messages.StopObservingBingoEventsReq,
                Messages.StopObservingBingoEventsRes> {
    @Override
    public CompletionStage<Messages.StopObservingBingoEventsRes> handle(
            BingoRoomSpot spot,
            PlayerActor actor,
            ZLinkMessageContext context,
            Messages.StopObservingBingoEventsReq request) {
        return CompletableFuture.completedFuture(spot.stopObserving(actor, request));
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/StopObservingBingoEventsHandler.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingSpotActorRequestHandler
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot
import systems.zlink.samples.kotlin.bingo.shared.contracts.StopObservingBingoEventsReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.StopObservingBingoEventsRes

class StopObservingBingoEventsHandler :
    ZLinkSuspendingSpotActorRequestHandler<
        BingoRoomSpot,
        PlayerActor,
        StopObservingBingoEventsReq,
        StopObservingBingoEventsRes,
    > {
    override suspend fun handle(
        spot: BingoRoomSpot,
        actor: PlayerActor,
        context: ZLinkMessageContext,
        request: StopObservingBingoEventsReq,
    ): StopObservingBingoEventsRes = spot.stopObserving(actor, request)
}
